Make protocol participant single-use

marsella opened the pull request:
Things this does:
- Updates protocol participant trait into two: one with external-facing methods and one with internal-only methods. Pulls a few methods that already existed into the traits.
- Genericizes `Participant` over `PublicProtocolParticipant`.
- Assigns a single session ID and protocol input to a `Participant` at generation time.
- Updates full-protocol test and e2e benchmarks to fully work with the new API!
- Removes network example from compilation with `autoexamples = false` (the network example does not use the new API).

Tasks that remain to complete
- Remove commented-out code (make decision on is-done discussion topic)

Things that would be better in separate issues:
- Update `PresignParticipant` to take a specific input, with focus on mapping from (new, presign) pids to public values from other parties. I think to do this correctly we'll need to change the output types of keygen and auxinfo, to be pairs of (pid, public info), instead of having pid be a field on the public info. #236.
- Add a `done?` method to `PublicProtocolParticipant` that queries local storage at most, or is just a field set at completion time. #160

marsella commented on src/protocol.rs:
```diff
@@ -121,145 +131,70 @@ impl Participant {
+                message.id(),
+                self.sid
+            );
+            return Err(InternalError::ProtocolError);
```
nit: this and the above "wrong message recipient" should be a "calling application error" instead.

marsella commented on src/protocol.rs:
```diff
@@ -121,145 +131,70 @@ impl Participant {
+            );
+            return Err(InternalError::ProtocolError);
+        }
+        match (message.message_type(), P::protocol_type()) {
```
I think this could be simplified: the match stays the same, but the body of the "acceptable" arms goes to nothing, since they're all the same. If this doesn't throw an error, then we call `process message` and handle the outcome identically for all `P`.

marsella commented on src/protocol.rs:
```diff
@@ -121,145 +131,70 @@ impl Participant {
+    /// Produce a message to signal to this participant that the protocol can
+    /// begin.
+    ///
+    /// Possible bug: previously, the presign version called
+    /// [`PresignParticipant::initialize_presign_message()`], which does
+    /// some additional record-keeping of the associated auxinfo and keygen
+    /// SIDs. Does that still need to be done?
```
note: no, not once we're single-sid. that map is used for maintaining a record of which keygen + auxinfo material to use for presign.

marsella commented on src/protocol.rs:
```diff
@@ -47,61 +53,25 @@ pub struct Participant {
-    /// Instantiate a new quorum of participants of a specified size. Random
-    /// identifiers are selected
-    #[instrument(skip_all, err(Debug))]
-    pub fn new_quorum<R: RngCore + CryptoRng>(
```
note: I moved this functionality to `ParticipantConfig::random_quorum`, since we want to reuse participant IDs among different `Participant` types. Now a caller can create the random config quorum and instantiate the `Participant`s themselves.

marsella commented on src/protocol.rs:
```diff
@@ -121,145 +91,56 @@ impl Participant {
-            keyshare_identifier,
-            identifier,
-        )
+    pub fn initialize_message(&self) -> Message {
```
note: this replaces three functions (`initialize_{keygen | auxinfo | presign}_message`).

marsella commented on src/protocol.rs:
```diff
@@ -561,9 +462,13 @@ mod tests {
     /// `presign_identifier` identifies a sub-session for pre-signing and
     /// signing.
-    fn is_presigning_done(quorum: &[Participant], presign_identifier: Identifier) -> Result<bool> {
+    fn is_presigning_done(
```
note: commented these out because of aforementioned rabbit hole, but they've been replaced by just checking that all participants have returned output.

marsella commented on src/protocol.rs:
```diff
@@ -641,81 +555,153 @@ mod tests {
+            .into_iter()
+            .map(|config| Participant::<KeygenParticipant>::from_config(&config, keygen_sid, ()))
+            .collect::<Vec<_>>();
+        let mut inboxes = HashMap::from_iter(
+            keygen_quorum
+                .iter()
+                .map(|p| (p.id, vec![]))
+                .collect::<Vec<_>>(),
+        );
```
note: actually, we can probably reuse the inboxes in this case, since they should be empty after a successful protocol run.

marsella commented on src/protocol.rs:
```diff
@@ -641,81 +555,153 @@ mod tests {
+        let _configs = ParticipantConfig::random_quorum(QUORUM_SIZE, &mut rng);
+
+        /*
+                // Prepare presign inputs: a pair of outputs from keygen and auxinfo, with the same participant ID for each one.
+                // Blocking on #236.
```
note: Formatting input for presign is why this is blocking on #236. I guess we could put it all into a `Storage` but I think that is not a useful intermediate state.

marsella commented on src/protocol.rs:
```diff
@@ -121,145 +91,56 @@ impl Participant {
+        Message::new(P::ready_type(), self.sid, self.id, self.id, &[])
     }
 
+    /* skipping this for now because I think it's a rabbit hole of a refactor
```
discussion topic: I commented these out because I didn't want to figure out how to check for done-ness, or to add #160 in this PR as well, but it turns out I don't need them any more because I can just check (in a good protocol run) whether all parties returned output. 

We probably still want to do #160 so we can check bad protocol runs (or e.g. that tests fail the expected way) but is there any objection to my removing this functionality now? I can update 160 with an "add public API method to check done-ness" so we remember to bring it back.

marsella replied:
@amaloz Do you have an opinion on this?

amaloz replied:
I'm all for removing this functionality now and keeping #160 around for bad protocol runs.

amaloz commented on src/participant.rs:
```diff
@@ -168,20 +168,26 @@ where
-    type Input;
+/// These are the public-facing methods that must be implemented for a given
+/// protocol.
+pub trait PublicProtocolParticipant {
```
Minor, but I'd prefer a renaming of `PublicProtocolParticipant` -> `ProtocolParticipant` and `ProtocolParticipant` -> `InnerProtocolParticipant` (or something like that). The reasoning is that the public API only exposes `PublicProtocolParticipant` so the `Public` seems redundant.

amaloz commented on src/participant.rs:
```diff
@@ -168,20 +168,26 @@ where
+    fn protocol_type() -> ProtocolType;
+
+    /// Create a new `PublicProtocolParticipant` from the given ids.
+    fn from_ids(
```
Should we rename this `new`?

amaloz commented on src/zkp/pimod.rs:
```diff
@@ -238,7 +238,7 @@ fn square_roots_mod_prime(n: &BigNumber, p: &BigNumber) -> Result<(BigNumber, Bi
         return Ok((r, neg_r));
     }
-    warn!("Could not find square roots modulo n");
+    //warn!("Could not find square roots modulo n");
```
Agreed. I noticed this in some of my testing as well, and did the same thing of commenting out this line.

marsella replied:
I will add it back in for now, but I updated #50.

amaloz commented on src/storage.rs:
```diff
@@ -46,8 +47,9 @@ impl<T: Storable> Storable for StorableIndex<T> {}
 
 #[derive(Serialize, Deserialize, Clone, Debug)]
-pub(crate) struct Storage(HashMap<Vec<u8>, Vec<u8>>);
+pub struct Storage(HashMap<Vec<u8>, Vec<u8>>);
```
Curious why this needs to be `pub`?

marsella replied:
I think this was to compile an intermediate state when `Presign` still took `Storage` as input. Probably not necessary any more!

amaloz commented on src/protocol.rs:
```diff
@@ -38,73 +27,44 @@ use tracing::{error, info, instrument, trace};
 /////////////////////
 
+#[derive(Debug)]
+pub enum ProtocolType {
```
Missing documentation.

marsella replied:
Updated. I'm not quite sold on this type right now, since the caller doesn't actually have to explicitly interact with it -- it's automatically instantiated based on what `ProtocolParticipant` type the caller uses to make a `Participant`. But I think we should leave it for now and just keep it in mind for potential refactoring once we've seen the example and so on.

amaloz commented on src/protocol.rs:
```diff
@@ -38,73 +27,44 @@ use tracing::{error, info, instrument, trace};
-    /// Participant subprotocol for handling presign messages
-    presign_participant: PresignParticipant,
+
+    sid: Identifier,
```
Missing documentation for these entries.

amaloz commented on src/protocol.rs:
```diff
@@ -38,73 +27,44 @@ use tracing::{error, info, instrument, trace};
-    /// Local storage for this participant to store finalized auxinfo, keygen,
-    /// and presign values. This storage is not responsible for storing
-    /// round-specific material.
-    main_storage: Storage,
```
:fireworks: ! Does this mean we can remove `Storage` entirely?

marsella replied:
I think so! I slapped an `#[allow(unused)]` on there at some point in this PR to get through an intermediary state, but I think we might be set to jettison it entirely. Will double-check.

marsella replied:
It's gone!!

amaloz commented on src/protocol.rs:
```diff
@@ -327,19 +202,17 @@ impl Participant {
-        presign_identifier: Identifier,
-        digest: sha2::Sha256,
-    ) -> Result<SignatureShare> {
+    pub fn sign(record: PresignRecord, digest: sha2::Sha256) -> Result<SignatureShare> {
```
Update associated documentation. Also, should this live in `Participant` anymore, or be moved to `PresignRecord`?

marsella replied:
I made #251 to move this to `PresignRecord`, yeah. The API is really awkward right now :) Omar had proposed having `Sign` be its own ProtocolParticipant, but I think we might not want that for now because there aren't any `Message`s in signing, it's just the local operation -- the calling application has to figure out where to send the signature shares once they get them (e.g. since it's often not the computation parties who are supposed to end up with the sig).

amaloz reviewed (changes requested):
I'm liking how this is shaping up! A couple of minor comments.

marsella commented on src/auxinfo/auxinfo_commit.rs:
```diff
@@ -52,7 +52,7 @@ impl Debug for AuxInfoDecommit {
             .field("rid", &"[redacted]")
             .field("u_i", &"[redacted]")
-            .field("public keys", &self.public_keys)
+            .field("public keys", &"[redacted -- tooooo long]") //self.public_keys)
```
Adding #?? to fingerprint this instead of printing the whole keys out.

marsella commented:
@amaloz This is ready for re-review! I addressed all your comments -- the changes are mostly doc tweaks and a bunch of dead code clean-up. 

I'm waiting to rebase until I can squash-and-rebase all at once. Preserving the commit diffs for now so you can actually see what changed.

marsella commented:
Also: I added #256 to add some tests about bad message routing.

amaloz commented on src/protocol.rs:
```diff
@@ -456,16 +219,40 @@ impl SignatureShare {
 
+#[cfg(test)]
+impl ParticipantConfig {
+    /// Get a list of `size` consistent `ParticipantConfig`s.
```
minor: wrap `ParticipantConfig` and `ParticipantIdentifier` in brackets

marsella commented on src/presign/participant.rs:
```diff
@@ -258,22 +264,9 @@ impl PresignParticipant {
     /// `auxinfo_identifier`, `keyshare_identifier` and `identifier` correspond
```
Good catch. In fact, this whole function is irrelevant. I don't know why clippy didn't notice that it's unused.

marsella commented on src/participant.rs:
```diff
@@ -168,20 +169,23 @@ where
-    fn local_storage_mut(&mut self) -> &mut LocalStorage;
-    fn id(&self) -> ParticipantIdentifier;
-    fn other_ids(&self) -> &Vec<ParticipantIdentifier>;
+    /// The `MessageType` of a "ready" message, signalling that a participant is
```
This is weird, actually. `MessageType` is a public type in a private module, so it's totally opaque to the user, but apparently the public trait can still return it. I rephrased the documentation to not reference the type itself (because the link breaks if I add brackets).
